### ISSUE TITLE
Alias completions with external completers does not work anymore (after #8123)

### ISSUE THREAD

**Issue by thomasgoulet** [nushell/nushell#8483]
### Describe the bug

I use fish as a completer with the exact same setup as [in the docs](https://www.nushell.sh/book/custom_completions.html#external-completions).

Before, in 0.76 I could `alias k = kubectl` and get automcompletion for the resource names from the fish completion file (`kubectl completion fish`)

In 0.77, this completion does not work anymore and I have to use the full command to get completion.

### How to reproduce

1. `alias k = kubectl`
2. `> k get pods <TAB>`

(0. Have kubectl installed and external completion setup)

### Expected behavior

Autocomplete as if I typed `kubectl get pods`.

### Screenshots

![image](https://user-images.githubusercontent.com/51253506/225660583-5ff53664-1733-4bdc-b312-8cd98c3dba9d.png)

(Not screenshotting expected behavior because I'm at work)

### Configuration

| key                | value                                                            |
| ------------------ | ---------------------------------------------------------------- |
| version            | 0.77.0                                                           |
| branch             | makepkg                                                          |
| commit_hash        | fd09609b44b53d41ac7929e90708cdeba7bfb571                         |
| build_os           | linux-x86_64                                                     |
| build_target       | x86_64-unknown-linux-gnu                                         |
| rust_version       | rustc 1.68.0 (2c8cc3432 2023-03-06) (Arch Linux rust 1:1.68.0-1) |
| cargo_version      | cargo 1.68.0                                                     |
| build_time         | 2023-03-15 09:27:56 +00:00                                       |
| build_rust_channel | release                                                          |
| features           | default, zip                                                     |
| installed_plugins  |                                                                  |

**Comment by sholderbach** [comment]
I think this is a subissue of the regression observed in #8471. So the changes to the aliases do not play nice with the completions yet.

**Comment by thomasgoulet** [comment]
> I think this is a subissue of the regression observed in #8471. So the changes to the aliases do not play nice with the completions yet.

The issue you mentioned is within 0.76.1, so the new alias changes were not in place.

**Comment by bartlibert** [comment]
I also have this issue and I noticed that aliases get expanded wrongly when passed to the external completer.
For example, I have an alias `alias gp = git push`.
I noticed parameters that are passed to the completer in this case are `gp push` instead of `git push`. 
So it seems the alias is expanded, but the first part is not replaced with the first part of the expanded alias.

I have the following (ugly) workaround in my config and this seems to work (for carapace):

```
let carapace_completer = {|spans|
    let has_alias = ($nu.scope.aliases | where name == $spans.0)
    let spans = (if not ($has_alias | is-empty) {
      # put the first word of the expanded alias first in the span
      $spans | skip 1 | prepend ($has_alias | get expansion | split words | get 0)
    } else { $spans })
    carapace $spans.0 nushell $spans | from json

```

**Comment by aserowy** [comment]
Heho,

any progress on this issue?

Kind regards
Alexander

**Comment by aserowy** [comment]
> I also have this issue and I noticed that aliases get expanded wrongly when passed to the external completer. For example, I have an alias `alias gp = git push`. I noticed parameters that are passed to the completer in this case are `gp push` instead of `git push`. So it seems the alias is expanded, but the first part is not replaced with the first part of the expanded alias.
> 
> I have the following (ugly) workaround in my config and this seems to work (for carapace):
> 
> ```
> let carapace_completer = {|spans|
>     let has_alias = ($nu.scope.aliases | where name == $spans.0)
>     let spans = (if not ($has_alias | is-empty) {
>       # put the first word of the expanded alias first in the span
>       $spans | skip 1 | prepend ($has_alias | get expansion | split words | get 0)
>     } else { $spans })
>     carapace $spans.0 nushell $spans | from json
> ```

Heho,

fyi. This snippet is not working anymore. In the cookbook is a running [workaround](https://www.nushell.sh/cookbook/external_completers.html#alias-completions).

Regards
Alexander

**Comment by inyutin** [comment]
Hello, the workaround doesn't work if alias use variable

Basically, if I have a alias like this:
```nushell
alias aws_test = aws --profile "admin-test"
```

Then carapace (external completer) would provide completions for aws as expected

However, if I define alias like this:
```nushell
let aws_admin_test_profile = "admin-test"
alias aws_test = aws --profile $aws_admin_test_profile
```

Then completions wouldn't work.

That's, I guess, is expected, because the workaround replaces alias with it's "expansion". So in the second case caparace tries to expand "aws --profile $aws_admin_test_profile" and fails to do so. 

Would be very nice to have a workaround for this case too, however I couldn't come up with one myself

**Comment by kaiphat** [comment]
I suppose it connects with `completions.algorithm = fuzzy`, because only in such case completion on one letter alias isn't working.

**Comment by kaiphat** [comment]
Lifehack:) Try to do two spaces after command, like this: `g  ` and then press tab

**Comment by WieeRd** [comment]
If there isn't a definitive 'fix' or planned refactoring about this, can the current workaround of expanding the aliases just be upstreamed / embedded to the Rust codebase itself? External completers could just receive expanded form in the first place. Or is there something making this hard to implement in the Rust src and can only be solved in `config.nu`?

**Comment by brglng** [comment]
I found the external completer now does not work with custom commands (`def`s) either.

**Comment by alecStewart1** [comment]
> If there isn't a definitive 'fix' or planned refactoring about this, can the current workaround of expanding the aliases just be upstreamed / embedded to the Rust codebase itself? External completers could just receive expanded form in the first place. Or is there something making this hard to implement in the Rust src and can only be solved in `config.nu`?

I'm gonna bump this while expanding on what @WieeRd said. In Fish Shell aliases are technically just a wrapper around fish's `function`:

> alias is a simple wrapper for the function builtin, which creates a function wrapping a command. It has similar syntax to POSIX shell alias. For other uses, it is recommended to define a [function](https://fishshell.com/docs/current/cmds/function.html).
> [ ... ]
> fish marks functions that have been created by alias by including the command used to create them in the function description.

And I assume that's one component of how fish's aliases can have the completions of the commands they wrap, or maybe it is *the* way.

I could be wrong, but maybe Nushell doesn't do this, and that an `alias` in Nushell is something other than a command defined with `def`? If it does, then I guess the bug is somewhere related to how `alias` wraps what's passed to it.

**EDIT:**

...or, as, you know, implied by what @aserowy posted: aliases aren't being expanded properly under the hood.

**Comment by simonborje** [comment]
I've been trying out ways to get the alias expanded before being sent to the external completers. Based on some very limited testing I found that the expansion seems to work after applying the changes that can be found here: https://github.com/simonborje/nushell/commit/4f71467508272601a9a7800e94407f178f7987a9.
I've only tested with a simple "external completer" that simply prints the arguments passed to it but it seems to work ok.

```
alias glo = git log --oneline

// Without patch
glo test<TAB>   =>   external completer called with ["glo" "log" "--oneline" "test"}

// With patch
glo test<TAB>   =>   external completer called with ["git" "log" "--oneline" "test"}
```

**Comment by thomas-goulet** [comment]
> I've been trying out ways to get the alias expanded before being sent to the external completers. Based on some very limited testing I found that the expansion seems to work after applying the changes that can be found here: [simonborje@4f71467](https://github.com/simonborje/nushell/commit/4f71467508272601a9a7800e94407f178f7987a9).`

Looks like a simple way to fix it! Maybe you could open a PR and we could see where it goes from there?

Personally, I've re-worked my workflow to dance around the issue and am mostly using custom completions for all commands I often use. I would still be interested in bringing back alias completion to my shell :)

**Comment by Malix-Labs** [comment]
The documentation is still referencing this issue as active : https://www.nushell.sh/cookbook/external_completers.html#alias-completions

**Comment by Malix-Labs** [comment]
See #17853 (which is related to this issue)
